src/utils/tracked-controls.js

```diff
@@ -46,41 +46,81 @@ module.exports.checkControllerPresentAndSetup = function (component, idPrefix, q
+  return !!findMatchingController(gamepads, null, idPrefix, queryObject.hand, targetControllerNumber);
+}
+
+module.exports.isControllerPresent = isControllerPresent;
+
+/**
+ * Walk through the given controllers to find any where the device ID equals filterIdExact, or startWith filterIdPrefix.
+ * A controller where this considered true is considered a 'match'.
+ *
+ * For each matching controller:
+ *   If filterHand is set, and the controller:
+ *     is handed, we further verify that controller.hand equals filterHand.
+ *     is unhanded (controller.hand is ''), we skip until we have found a number of matching controllers that equals targetControllerNumber
+ *   If filterHand is not set, we skip until we have found the nth matching controller, where n equals targetControllerNumber
+ *
+ * The method should be called with one of: [filterIdExact, filterIdPrefix] AND one or both of: [filterHand, targetControllerNumber]
+ *
+ * @param {object} controllers - Array of gamepads to search
+ * @param {string} filterIdExact - If set, used to find controllers with id === this value
+ * @param {string} filterIdPrefix - If set, used to find controllers with id startsWith this value
+ * @param {object} filterHand - If set, further filters controllers with matching 'hand' property
+ * @param {object} targetControllerNumber - Find the nth matching controller, where n equals targetControllerNumber. defaults to 0.
+ */
+function findMatchingController (controllers, filterIdExact, filterIdPrefix, filterHand, targetControllerNumber) {
+  var controller;
+  var i;
+  var matchingControllerOccurence = 0;
+  targetControllerNumber = targetControllerNumber || 0;
+
+  for (i = 0; i < controllers.length; i++) {
+    controller = controllers[i];
+    // Determine if the controller ID matches our criteria
+    if (filterIdPrefix && controller.id.indexOf(filterIdPrefix) === -1) continue;
+    if (!filterIdPrefix && controller.id !== filterIdExact) continue;
+
+    if (filterHand) {
+      if (filterHand === controller.hand) {
+        // If the component requests a specific hand and found a matching one, we ignore the
+        // targetControllerNumber requirement and early exit.
+        return controller;
+      } else if (controller.hand) {
+        continue;
+      }
+      // If we reach here, the controller is unhanded - check against targetControllerNumber
     }
-    if (isPresent && queryObject.index) {
-      // Need to use count of gamepads with idPrefix.
-      isPresent = index === queryObject.index;
+    // The controller is unhanded, or we are looking for the nth occurence of a matching controller (n equals targetControllerNumber).
+    if (matchingControllerOccurence === targetControllerNumber) {
+      return controller;
     }
-    if (isPresent) { break; }
-    // Update count of gamepads with idPrefix.
-    if (isPrefixMatch) { index++; }
+    ++matchingControllerOccurence;
   }
-
-  return isPresent;
+  return null;
```

Comment:
probably better `return undefined` for consistency with other functions across the code base